crates/bevy_color/src/color_gradient.rs

```diff
@@ -0,0 +1,104 @@
+use crate::Mix;
+use bevy_math::curve::{cores::EvenCoreError, Curve, Interval, SampleCurve};
+
+/// A curve whose samples are defined by a collection of colors.
+#[derive(Clone, Debug)]
+pub struct ColorCurve<T: Mix + Clone, I> {
+    curve: SampleCurve<T, I>,
+}
+
+impl<T, I> ColorCurve<T, I>
+where
+    T: Mix + Clone,
+{
+    /// Create a new [`ColorCurve`] from a collection of [mixable] types and a mixing function. The
+    /// domain of this curve will always be `[0.0, len - 1]` where `len` is the amount of mixable
+    /// objects in the collection.
+    ///
+    /// This fails if there's not at least two mixable things in the collection.
+    ///
+    /// [mixable]: `Mix`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_color::palettes::basic::*;
+    /// # use bevy_color::Mix;
+    /// # use bevy_color::Srgba;
+    /// # use bevy_color::ColorCurve;
+    /// # use bevy_math::curve::Interval;
+    /// # use bevy_math::curve::Curve;
+    /// let broken = ColorCurve::new([RED], Srgba::mix);
+    /// assert!(broken.is_err());
+    /// let gradient = ColorCurve::new([RED, GREEN, BLUE], Srgba::mix);
+    /// assert!(gradient.is_ok());
+    /// assert_eq!(gradient.unwrap().domain(), Interval::new(0.0, 2.0).unwrap());
+    /// ```
+    pub fn new(
+        colors: impl IntoIterator<Item = impl Into<T>>,
+        interpolation: I,
+    ) -> Result<Self, EvenCoreError>
+    where
+        I: Fn(&T, &T, f32) -> T,
+    {
+        let colors = colors.into_iter().map(Into::into).collect::<Vec<_>>();
+        Interval::new(0.0, colors.len().saturating_sub(1) as f32)
+            .map_err(|_| EvenCoreError::NotEnoughSamples {
+                samples: colors.len(),
+            })
+            .and_then(|domain| SampleCurve::new(domain, colors, interpolation))
+            .map(|curve| Self { curve })
+    }
+}
+
+impl<T, F> Curve<T> for ColorCurve<T, F>
+where
+    T: Mix + Clone,
+    F: Fn(&T, &T, f32) -> T,
+{
+    fn domain(&self) -> Interval {
+        self.curve.domain()
+    }
+
+    fn sample_unchecked(&self, t: f32) -> T {
+        self.curve.sample_unchecked(t)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::palettes::basic;
+    use crate::Srgba;
+
+    #[test]
+    fn test_color_curve() {
+        let broken = ColorCurve::new([basic::RED], Srgba::mix);
+        assert!(broken.is_err());
+
+        let gradient = [basic::RED, basic::LIME, basic::BLUE];
+        let curve = ColorCurve::new(gradient, Srgba::mix).unwrap();
```

Comment:
```suggestion
        let broken = ColorCurve::new([basic::RED]);
        assert!(broken.is_err());

        let gradient = [basic::RED, basic::LIME, basic::BLUE];
        let curve = ColorCurve::new(gradient).unwrap();
```

And after all that we get the cleaner API we hoped for!